### ISSUE TITLE
Make 'self' a direct attribute of a resource object

### ISSUE THREAD

**Issue by bintoro** [json-api/json-api#370]
As it stands, the workings of `self` are inconsistent.

Inside a _link object_, `self` indicates the URL of the thing represented by the **object itself**. The `self` member appears together with the object's `id` and `type`.

Inside a _links object_, `self` points to the URL of the thing represented by the **parent object**. Given the previous observation, this is illogical.

This may not seem like a huge deal, but there are a couple of another points here too:
- In JSON API, the `links` member is intended as a container for _relationships_. It seems like a needless complication to stuff an extra key there just for the fun of it, when it has nothing to do with relationships. This relates to my ongoing effort to put some sense in the link/links/linked/related/included conceptual mess (#380).
- Clients may need to look up canonical URLs on the `self` members of related resource objects. Assuming you want to exclude the actual resource attributes, how do you request the inclusion of `self` when it's neither a relationship nor an attribute?

I propose to move the offending `self` member one level up to the context it actually belongs to:

```
{
  "data": [{
    "type": "articles",
    "id": "1",
    "self": "http://example.com/articles/1",           <---
    "title": "JSON API paints my bikeshed!",
    "links": {
      "author": {
        "self": "http://example.com/articles/1/links/author",
        "resource": "http://example.com/articles/1/author",
        "type": "people",
        "id": "9"
      },
      "comments": {
        "self": "http://example.com/articles/1/links/comments",
        "resource": "http://example.com/articles/1/comments",
        "type": "comments",
        "ids": ["5", "12"]
      }
    }
  }]
}
```

Now `self`, `id`, and `type` are aligned both at the resource level and the link level.

There's no reason why `self` needs to be inside `links`. I presume this idea came from HAL, but the ways in which `links` is used in HAL and JSON API are not really analogous.

**Comment by dgeb** [comment]
@wycats and I aren't seeing this as an inconsistency. Will comment more later.

**Comment by bintoro** [comment]
@dgeb, understood.

Canonical URL troubles would be better addressed directly in the link object anyway (#379). Closing because this is just an annoyance, not a huge defect.

But if #379 doesn't happen (the two-key solution), and `self` can only be delivered within`links`, then JSON API isn't going to be a very elegant format for applications that use canonical URLs for linking.
